[Http] General design, structure and convention guidelines (#226558)
elastic/kibana: dev_docs/contributing/kibana_http_api_design_guidelines.mdx

---
id: kibHttpApiGuidelines
slug: /kibana-dev-docs/contributing/http-api-guidelines
title: Guidelines for HTTP API design in Kibana
description: Best practices and guidelines for designing consistent, user-friendly HTTP APIs in Kibana
date: 2025-07-02
tags: ['kibana', 'contributor', 'dev', 'http', 'api', 'rest', 'design']
---

[Click here](#structure-and-conventions) to skip to the structure and conventions guides.

## Design principles

### Consistency

Consistency is key to a great API-user experience. APIs should be consistent in their naming, structure, and behavior. Follow established patterns when available. If no pattern exists, consider creating one paying close attention to:

1. **Consistent naming**: Use the same suffix for IDs (e.g., `space_id`, `dashboard_id`, etc.). APIs with clear, descriptive field names help self-document how it should be used.
2. **Consistent structure**: Use the same structure for referencing resources in requests and responses in all related APIs.
3. **Consistent validation**: Apply the same validation rules across all APIs.

> 4. **Consistent error messages**: Use the same error messages for the same errors.

Reviewer: Are we speaking about identical errors & matching messages for different endpoints? For example, 

```
GET /api/foos/{id}
GET /api/foos
```

GET single object throws `404` with error message `not found`.
GET list of objects throws `404` with error message `not found`.

Author: I think we are talking more about the structure of error messages. I'll clarify this point and refer to later in the guide.

This principle is embodied in the rest of this document, so read on for examples!

### Make a great API-user experience

Create a great _API_ user experience, as opposed to optimizing a single API for _UI_ use cases as well. Front-ends have very different needs and constraints from public HTTP API clients. Our public HTTP APIs should be primitives that are simple to use and compose.

For example, a UI might want to optimize it's network requests by bundling multiple requests into a single request, but an API client would be better designed to de-optimize and simplify the requests:

A view or component in the UI might do the following for some, UI-specific purpose:
```
GET /internal/some_related_resources_for_this_page?ids_only=true&page=1
# Which returns:
{
  "a_ids": ["1", "2"],
  "b_ids": ["2"]
}
```

While API clients are better designed as:
```
# Request 1
GET /api/resource_as?page=1
{
  "items": [{ id: "1", "name": "Resource A 1" }, { id: "2", "name": "Resource A 2" }],
  "page": 1,
  "total": 2
}
# Request 2
GET /api/resource_bs?page=1

> {

Reviewer: I'll let this fly for now to unblock the PR but we have to follow up to explicitly show that `GET` doesn't accept a body.

 - Add a new required property
 - Set an existing **optional** property as required
 - Add or delete a security requirement
 - In some cases, changing a default
 - In some cases, _behavioural_ changes might form part of your contract

#### Non-breaking changes to HTTP APIs include
 - Adding a new **optional** request parameter
 - Expanding the types of input parameters, going from `string` -> `string | string[]` in requests
 - Adding a new **optional** response property
 - Relaxing request validation requirements: e.g. going from `schema.string({ minLength: 10 })` -> `schema.string()`

<DocCallOut title="Do not break clients!">
  A public HTTP API should never cause a client to break. Not without a long deprecation period and a ready alternative.

  Even `Technical Preview` HTTP APIs should consider graceful paths for changes when possible.

  Linus Torvalds famously said of Kernel development: "WE DO NOT BREAK USERSPACE!" (along with some other expletives). We should adopt this kind of rigor and empathy when working with our _public_ HTTP APIs and always prioritize finding ways to avoid breaking our APIs.
</DocCallOut>

## Internal vs Public APIs

> Public HTTP APIs are exposed to users. They are also used by plugins and external tools. They are the focus of this document.

Author: Good idea! I'll move this up.

* `DELETE` always results in the same _server state_ (returns 404 after the first request)
* `PATCH` always results in the same _server state_ for the same input! For example: `PATCH /api/coolstore/order/{id} { expected_delivery: "ASAP!" }` will result in the same state.

However...

* `POST` should result in a different _server state_. For example: `POST /api/coolstore/order` will result in a new order each time!

**POST is (mostly) for creating a resource**

Sometimes `POST` can be used for an action against a resource or for actions like a `GET` but with a body. In this case, `POST` may be both side-effect free and idempotent! In most cases it will not be because your request will open a PIT: see [pagination and sorting](#pagination-and-sorting).

If you expect you need to support a lot of input for searching or listing resources it might be good to reserve `POST` for this purposes.

**GET does not accept a body**

`GET /api/my_domain/my_resources/{id}` ✅

`GET /api/my_domain/my_resources/{id} { body: { params } }` ❌

```
POST /api/my_domain/my_resources
{
  search_after: "<a_really_long_id>",
  size: 10000
}
```

> **PUT or PATCH are for updating a resource**

Reviewer: ```suggestion
**PUT and PATCH are for updating a resource**
```

> `PUT` is a bit special in that it should _create_ the resource if it does not exist. It should expect the **full** resource in the body of the call.

Reviewer: ```suggestion
`PUT` is a bit special in that it should _create_ the resource if it does not exist and expect the **full** resource in the body of the call.
```

`PATCH` is used for partial updates and should never create a resource that does not already exist. It should expect a **partial** resource in the body of the call.

**DELETE is for deleting a resource**

Delete can return a simple response instead of the full resource:

```
DELETE /api/my_domain/my_resources/{id}
# -> 204 No content
```

**Stick to simple methods**

> `GET`, `POST`, `PUT`, `PATCH`, `DELETE` should cover almost all cases. If you are considering `HEAD` or some other method make sure you have a good justification for your HTTP API.

Reviewer: ```suggestion
`GET`, `POST`, `PUT`, `PATCH`, `DELETE` are often enough to cover almost all cases. If you are considering `HEAD` or some other method make sure you have a good justification for your HTTP API.
```

Reduces use of "should"

### Path and query parameters

> **Should use snake case**

Reviewer: Is using snake case a suggestion or a must have?

"Should (as suggestion) use snake case" -> "Prefer snake case"

"Should (as must-have) use snake case" -> "Use snake case"

Reviewer: We could leave this as it is for now and follow up with changing it to `Must use snake case`

`/api/my_api/{my_id}` ✅

`/api/my_api/{myId}` ❌

`/api/my_api?snake_case=true` ✅

`/api/my_api?camelCase=false` ❌

> **BEWARE: path or query parameters should not expect values of unknown length**

Reviewer: ```suggestion
**BEWARE: path and query parameters should not expect values of unknown length**
```

Accepting very long strings (in excess of 200 characters per parameter) in path or query parameters can cause issues for HTTP servers that limit the byte length of certain parts of requests. HTTP servers may limit request header sections to as little as 4096 bytes!

Use a body parameter for long strings instead and use validation to limit a path param's length if you have an idea of max length.

### Bodies

**Should contain snake case keys**

```
POST /api/my_domain/my_api
{
  "snake_case": true
}
```

<DocCallOut title="Snakes and camels">

>   In Kibana's TypeScript code it is considered a formatting issue to use snake case variable names. This clashes with the convenient [object destructuring assignments](https://www.typescriptlang.org/docs/handbook/variable-declarations.html#object-destructuring) that are common in throughout the codebase.

Reviewer: ```suggestion
  In Kibana's TypeScript code it is considered a formatting issue to use snake case variable names. This clashes with the convenient [object destructuring assignments](https://www.typescriptlang.org/docs/handbook/variable-declarations.html#object-destructuring) that are common throughout the codebase.
```

**Resource shapes should be consistent**

`GET`, `PUT`, `POST` should return the same shape of data for the same resource.

`GET /api/my_domain/my_resources/{id}` <br />
=> <br />
`{ "id": "1", "name": "My resource" }`<br /><br />

`POST /api/my_domain/my_resources/{id} { "id": "1", "name": "My resource" }`<br />
=> <br />
`{ "id": "1", "name": "My resource" }`<br /><br />

`GET /api/my_domain/my_resources` <br />
=> <br />
`{ items: [{ "id": "1", "name": "My resource" }], page: 1, size: 10, total: 100 }`

See the section on [data modelling](#data-modeling) for more guidance.

### Defaults

**Should be used to promote ease of use**

Choose sensible defaults. When uncertain, ask API callers to make informed decisions based on [documentation](#documentation).

**Should not be changed lightly**

Changing the value of a default may, in some cases, have a devastating result similar to a breaking change.

### Validation

**Runtime validation should be as narrow as feasible**

`schema.object({ id: schema.string({ minLength: 32, maxLength: 32 }) })` ✅

`schema.object({ id: schema.string() })` ❌

> **It is easier to relax requirements than tighten them**

Reviewer: ```suggestion
**It is easier to relax requirements than tighten them up**
```
